CODING_GUIDELINES.md

```diff
@@ -1,37 +1,186 @@
+
+- A private property is hidden behind a getter (and maybe a setter) that has the same name:
+```typescript
+private _open: boolean;
+public get open(): boolean;
+public set open(value: boolean);
+
+private _openChange: Subject<boolean>;
+public get openChange(): Observable<boolean>;
+```
+- A property or method needs to be public for integration purposes but should not be considered public by the consumer:
+```typescript
+@ContentChildren(Something) _children: QueryList<Something>;
+_internalMethodCalledFromAnotherClass()
+```
+
+
+### Angular
+#### Components and directives selector
+All Angular selectors are prefixed with `clr`, then follow the typical `snake_case` or `camelCase` depending on whether they're element or attribute selectors:
+```
+clr-component
+[clrDirective]
+```
+When larger components have sub-components, the sub-component should use these selectors:
+```
+clr-component-subcomponent
+[clrComponentSubdirective]
+```
+This is slightly verbose, but will prevent conflicts between components. This is a case where abbreviations are acceptable to reduce verbosity.
+
+#### Inputs and outputs
+To avoid conflict with non-Clarity components, inputs and outputs of our directives and components should be prefixed with `clr`:
+```
+[clrInput]
+(clrInputChange)
+(clrOutput)
+```
+
+#### Components and directives class names
+Components and directives are exported Typescript classes, so they respect the earlier convention and are prefixed with `Clr`. Contrary to the recommendation on the Angular style guide, they do not have any `Component` or `Directive` suffix, in order to allow us to switch from one to the other without breaking changes. For instance:
+```
+ClrWizard (component)
+ClrIfOpen (structural directive)
+ClrLoading (directive)
+```
+
+#### Components and directives class names
+Once again, providers are exported Typescript classes so they are prefixed. To avoid conflicts with similar component names, they are suffixed with `Service`:
+```
+ClrSomeService
+ClrIfOpenService
+```
+
+#### Modules
+Prefixed with `Clr` as usual, and suffixed with `Module`:
+```
+[TODO: main Clarity module] ClrModule / ClarityModule / ClrTheOneModule ???
+ClrWizardModule
+ClrFormsModule
+```
+
+
+## Static UI
+
+- Remove unused HTML wrappers that might be left from previous prototypes of your code.
+- Avoid styling elements themselves, use CSS classes.
+- Keep your CSS selectors as flat as possible, try not to exceed 2 levels (e.g. `.parent .child`).
+- Do not hardcode colors, use the SCSS variables provided in `color/_variables.color.scss`.
+- Only use `rem` and `%` sizes. The only exception is `1px` borders that should remain 1px even on larger font sizes.
+- All of your SCSS styles should be wrapped in an `@include exports('XXX.clarity') { ... }`, to avoid producing duplicates in the deliverable CSS.
+- The baseline should always be respected to ensure proper vertical rhythm.
```

Comment:
> Changes should respect the 6px baseline when implmenting elements with known heights to ensure vertical rhythm (e.g. the line height of text is 12px, 18px, 24px, etc).

Reply:
I didn't want to put a hard pixel value here, because some people actually change our baseline. All we say is line heights are supposed to be multiples of our baseline, whatever that baseline is. By the way, ours is 24px, not 6px. The smaller pixel heights are accepted only because you have more than one of them and they add up to a multiple of 24px at the end.